### ISSUE TITLE
belongs_to with touch: true doesn't touch associated object on destroy

### ISSUE THREAD

**Issue by Hirurg103** [rails/rails#36751]
### Steps to reproduce
```ruby
begin require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "http://rubygems.org"
  gem "activerecord", "5.2.3"
  gem 'sqlite3', '< 1.4'
  gem "pry"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do

  create_table "posts", force: :cascade do |t|
    t.string "title", null: false
    t.text "body", null: false
    t.integer "touched_count", default: 0
  end

  create_table "comments", force: :cascade do |t|
    t.integer "post_id", null: false
    t.text "text", null: false
  end

end

class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
end

class Comment < ApplicationRecord
  belongs_to :post, touch: true

  #after_destroy { post.touch }
end

class Post < ApplicationRecord
  has_many :comments, dependent: :destroy

  after_touch { increment!(:touched_count) }
end


class IssueWithTouchInAssociationsInRails_5_2_Test < Minitest::Test

  def setup
    @post = Post.create!(title: "Issue", body: "Issue with touch in assosiations")

    @first_comment = Comment.create!(post_id: @post.id, text: "First comment")
    @second_comment = Comment.create!(post_id: @post.id, text: "Second comment")
  end

  def teardown
    @post.destroy
  end

  def test_touch_for_belongs_to_association
    @first_comment.destroy
    @second_comment.destroy

    @post.reload

    assert_equal 2, @post.touched_count
  end

end

test = IssueWithTouchInAssociationsInRails_5_2_Test.new('Touch error')

test.setup
test.test_touch_for_belongs_to_association
test.teardown
```
### Expected behavior
The `after_touch` callback on the post model should be called when its comment gets destroyed and increment `touched_count` attribute.

### Actual behavior
Post is not touched when its comment gets destroyed.

### System configuration
**Rails version**:: 5.2.0, 5.2.1, 5.2.2, 5.2.3

**Ruby version**:: 2.4.1p111

Related issues: #29078, #30842

**Comment by chukitow** [comment]
@Hirurg103  I think this is not a bug, just a bad implementation.

Documentation says:
```
:touch

If true, the associated object will be touched (the updated_at/on attributes set to current time) when this record is either saved or destroyed.
```

`after_touch` callback is not called in your example due to the Post model, you need to include a timestamp on your migration in order to make it work.

```ruby
  create_table "posts", force: :cascade do |t|
    t.string "title", null: false
    t.text "body", null: false
    t.integer "touched_count", default: 0
    t.timestamps
  end
```

But your test will still failing given it expects `touched_count` to be 2 instead of 4.  the `after_touch` callback is being triggered 4 times. (twice during comment creation and the twice when destroying the comments)

**Comment by Hirurg103** [comment]
Hi @chukitow , thank for your comment! Yes, I agree, I made an implementation mistake that didn't add timestamps to the `posts` table. 

I tried to simplify steps to reproduce that's why I replaced `HABTM` with `belogs_to` association. Steps to reproduce in my real case should have `HABTM` association with counter cache:

```ruby
begin require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "http://rubygems.org"
  gem "activerecord", "5.2.0"
  gem 'sqlite3', '< 1.4'
  gem "pry"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do

  create_table "notes", force: :cascade do |t|
    t.text "text", null: false
    t.integer "touched_count", default: 0
    t.integer "notes_count", default: 0
    t.timestamps
  end

  create_table "labellings", force: :cascade do |t|
    t.integer "note_id", null: false
    t.integer "label_id", null: false
  end

  create_table "labels", force: :cascade do |t|
    t.text "text", null: false
    t.timestamps
  end

end

class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
end

class Note < ApplicationRecord
  has_many :labellings, dependent: :destroy
  has_many :labels, through: :labellings

  after_touch { increment!(:touched_count) }
end

class Labelling < ApplicationRecord
  belongs_to :note, touch: true, counter_cache: :notes_count

  belongs_to :label

  # after_destroy { note.touch }
end

class Label < ApplicationRecord
  has_many :labellings, dependent: :destroy
end


class IssueWithTouchInAssociationsInRails_5_2_Test < Minitest::Test

  def setup
    @note = Note.create!(text: "Issue with touch in assosiations")

    @first_label = Label.create!(text: "Error")
    @second_label = Label.create!(text: "Bug")

    @note.labels = [@first_label, @second_label]
  end

  def teardown
    @note.destroy
    @first_label.destroy
    @second_label.destroy
  end

  def test_touch_for_has_many_through_association
    @note.reload

    assert_equal 2, @note.touched_count

    @note.labels.destroy_all
    @note.reload

    assert_equal 4, @note.touched_count
  end

end

test = IssueWithTouchInAssociationsInRails_5_2_Test.new('Touch error')

test.setup
test.test_touch_for_has_many_through_association
test.teardown
```

This bug is reproducible in Rails `5.2.0`, but looks like it was fixed in Rails `5.2.1`
